### PR TITLE
Make MongoDB health check a retryable operation

### PR DESCRIPTION
**Pull request by smnsht**
**What this PR does / why we need it**:
 Make MongoDB health check a retryable operation #2232 

**Which issue(s) this PR fixes**:
In some cases, the check can return a temporary not-ok status, while actually operations would still succeed without errors.

Please reference the issue this PR will close: #2232 

**Special notes for your reviewer**:
I am not sure if the project AspNetCore.Diagnostics.HealthChecks maintaners wants to rely on MongoDB C#/.Net driver team in the matter of converting ping command to be retryable, so I would like to propose retry logic implemented in the health check itself.

**Does this PR introduce a user-facing change?**:
No

Please make sure you've completed the relevant tasks for this PR, out of the following list:

- [*] Code compiles correctly
- [*] Created/updated tests
- [*] Unit tests passing
- [*] End-to-end tests passing
- [ ] Extended the documentation
- [ ] Provided sample for the feature

### CODE REVIEW

File: src/HealthChecks.MongoDb/MongoDbHealthCheck.cs [hunk 1]
```diff
@@ -46,10 +48,30 @@ public async Task<HealthCheckResult> CheckHealthAsync(HealthCheckContext context
                 // this you can check a specified database.
                 // Related with issue #43 and #617
 
-                await mongoClient
-                    .GetDatabase(_specifiedDatabase)
-                    .RunCommandAsync(_command, cancellationToken: cancellationToken)
-                    .ConfigureAwait(false);
+                for (int attempt = 1; attempt <= MAX_PING_ATTEMPTS; attempt++)
```
Comment (WeihanLi) [inline src/HealthChecks.MongoDb/MongoDbHealthCheck.cs:51]: Should we consider configure a delay between retry and the max retry count

Comment (smnsht) [reply src/HealthChecks.MongoDb/MongoDbHealthCheck.cs:51]: @WeihanLi 

I have been thinking about this.

Seems adding a delay between retries won't give any benefit.

If a connection attempt fails due to networking problems, the operation will take a long time to fail. Practically - could take 20-30 seconds, I guess it depends on Mongo client timeout settings, which we don't know.

Let's examine a simple case: MAX_PING_ATTEMPTS = 3, and Mongo client has a timeout of 30 seconds. In this case single check will take ~90 seconds. A second attempt is made after ~30 seconds, and the last attempt is made after ~1minute.

​If a check failed due to permissions/missing collection, etc... - delay between the checks does not have any benefit.

**Review by rohitvipin — APPROVED** [review]

**Review by adamsitnik — COMMENTED** [review]
> Please reference the issue this PR will close: #2232

Please correct me if I am wrong: Mongo DB driver performs 1 retry for most of the operations, but it does not for the method that our heatlh check is using.

The suggested solution is to add such retry on our side.

> In some cases, the check can return a temporary not-ok status, while actually operations would still succeed without errors.

To me it sounds like "by design" (the health check was for some reason unable to perform the ping), but I would prefer to ask @davidfowl and @damianedwards who architect ASP.NET and should be able to tell us what is the right approach in terms of retries performed by the health checks.

**Comment by DamianEdwards** [comment]
My initial thought is that the health check should use whatever resiliency strategy the app would regularly use so that the health check more accurately represents the app's ability to connect to the relevant resource.

File: src/HealthChecks.MongoDb/MongoDbHealthCheck.cs (hunk 1 above)
Comment (adamsitnik) [inline src/HealthChecks.MongoDb/MongoDbHealthCheck.cs:51]: Let's add a comment to explain why we are doing that:



```suggestion
                // For most operations where it is possible, the MongoDB driver itself will retry exactly once
                // to cover switches in the primary and temporary short term network outages.
                // Due to the RunCommand being a lower level function, according to the spec (https://github.com/mongodb/specifications/blob/master/source/run-command/run-command.rst#retryability)
                // for it, it is not retryable and this extends to the ping.
                for (int attempt = 1; attempt <= MAX_PING_ATTEMPTS; attempt++)

```

**Review by adamsitnik — APPROVED** [review]
LGTM, thank you for your contribution @smnsht !

**Pull request merged**